rectools/visuals/metrics_app.py

```diff
@@ -85,62 +96,160 @@ def construct(
         """
-        return cls(models_metrics, show_legend, auto_display, layout_kwargs)
+        cls._validate_models_metrics_base(models_metrics)
+        cls._validate_models_metrics_split(models_metrics)
+        cls._validate_models_metrics_names(models_metrics)
+        if models_metadata is None:
+            models_metadata = models_metrics[Columns.Model].drop_duplicates().to_frame()
+        cls._validate_models_metadata(models_metadata)
 
-    @property
-    @lru_cache
-    def metric_names(self) -> tp.List[str]:
-        """List of metric column names from the `models_metrics`."""
-        non_metric_columns = {Columns.Split, Columns.Model}
-        return [col for col in self.models_metrics.columns if col not in non_metric_columns]
+        merged_data = models_metrics.merge(models_metadata, on=Columns.Model, how="left")
+        merged_data[Columns.Model] = merged_data[Columns.Model].str.replace(" ", "_")
+
+        metric_names = [col for col in models_metrics.columns if col not in {Columns.Split, Columns.Model}]
+        meta_names = [col for col in models_metadata.columns if col != Columns.Model]
+
+        return cls(merged_data, metric_names, meta_names, show_legend, auto_display, scatter_kwargs)
 
     @property
     @lru_cache
     def model_names(self) -> tp.List[str]:
         """Sorted list of model names from `models_metrics`."""
-        return sorted(self.models_metrics[Columns.Model].unique())
+        return sorted(self.data[Columns.Model].unique())
 
     @property
-    def fold_ids(self) -> tp.List[int]:
+    @lru_cache
+    def fold_ids(self) -> tp.Optional[tp.List[int]]:
         """Sorted list of fold identifiers from the `models_metrics`."""
-        return sorted(self.models_metrics[Columns.Split].unique())
+        if Columns.Split in self.data.columns:
+            return sorted(self.data[Columns.Split].unique())
+        return None
 
-    def _validate_input_data(self) -> None:
-        if not isinstance(self.models_metrics, pd.DataFrame):
+    @staticmethod
+    def _validate_models_metrics_base(models_metrics: pd.DataFrame) -> None:
+        if not isinstance(models_metrics, pd.DataFrame):
             raise ValueError("Incorrect input type. `metrics_data` should be a DataFrame")
-        if Columns.Split not in self.models_metrics.columns:
-            raise KeyError("Missing `Split` column in `metrics_data` DataFrame")
-        if Columns.Model not in self.models_metrics.columns:
-            raise KeyError("Missing `Model`` column in `metrics_data` DataFrame")
-        if len(self.models_metrics.columns) < 3:
+        if Columns.Model not in models_metrics.columns:
+            raise KeyError("Missing `Model` column in `metrics_data` DataFrame")
+        if len([item for item in models_metrics.columns if item not in {Columns.Model, Columns.Split}]) < 1:
             raise KeyError("`metrics_data` DataFrame assumed to have at least one metric column")
+        if models_metrics.isnull().values.any():
```

Comment:
Do we need this for any case apart from dq check? Will nan values break anything from our functionality?
If they don't break anything, I think we can drop this check. From what I have tried, model with nan metrics remains in the legend but doesn't show on the chart if metric with nan value is selected. Which is good I think. What do you think?
(We shill need to check null values in Model and Split columns)

Reply:
Actually, no, we can safely keep these NaN values for metrics. I need to check how pandas handles these NaNs when computing average fold values, but as far as I can see, this is the only potential issue.

Reply:
Done